code/network/lib/Loot/Network/ZMQ/Adapter.hs

```diff
@@ -0,0 +1,182 @@
+{-# LANGUAGE BangPatterns #-}
+
+-- | Provides a poll-like methods to compose STM actions and polls on
+-- ZMQ sockets.
+
+module Loot.Network.ZMQ.Adapter
+       ( orElseMulti
+       , threadWaitReadSTMLong
+       , socketWaitReadSTMLong
+       , canReceive
+       ) where
+
+import Control.Concurrent (forkIO, threadDelay)
+import Control.Concurrent.STM.TQueue (TQueue, newTQueueIO, readTQueue, writeTQueue)
+import Control.Monad.STM (STM, orElse, retry)
+import Data.List.NonEmpty as NE
+import qualified GHC.Conc.IO as G
+import qualified GHC.Conc.Sync as Sync
+import System.Posix.Types (Fd)
+import qualified System.ZMQ4 as Z
+
+import Loot.Network.Utils (tMeasureIO, whileM)
+
+----------------------------------------------------------------------------
+-- Functions to export
+----------------------------------------------------------------------------
+
+-- | Sequential 'orElse' for nonempty list of STM actions.
+orElseMulti :: NonEmpty (STM a) -> STM a
```

Comment:
```haskell
orElseMulti = foldr1 orElse
```

Reply:
Right, NE is foldable too. Ok, but this method will be deleted anyway.